// @hashgraph/sdk/lib/file/FileId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _EntityId = _interopRequireWildcard(require("../EntityId.cjs"));

var proto = _interopRequireWildcard(require("@hashgraph/proto"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @typedef {import("long")} Long
 */

/**
 * The ID for a crypto-currency file on Hedera.
 *
 * @augments {EntityId<proto.IFileID>}
 */
class FileId extends _EntityId.default {
  /**
   * @param {number | Long | import("../EntityId.js").IEntityId} properties
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(properties, realm, num) {
    super(properties, realm, num);
  }
  /**
   * @param {string} text
   * @returns {FileId}
   */


  static fromString(text) {
    return new FileId(...(0, _EntityId.fromString)(text));
  }
  /**
   * @internal
   * @param {proto.IFileID} id
   * @returns {FileId}
   */


  static _fromProtobuf(id) {
    return new FileId({
      shard: id.shardNum != null ? id.shardNum : 0,
      realm: id.realmNum != null ? id.realmNum : 0,
      num: id.fileNum != null ? id.fileNum : 0
    });
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {FileId}
   */


  static fromBytes(bytes) {
    return FileId._fromProtobuf(proto.FileID.decode(bytes));
  }
  /**
   * @override
   * @internal
   * @returns {proto.IFileID}
   */


  _toProtobuf() {
    return {
      fileNum: this.num,
      shardNum: this.shard,
      realmNum: this.realm
    };
  }
  /**
   * @returns {Uint8Array}
   */


  toBytes() {
    return proto.FileID.encode(this._toProtobuf()).finish();
  }

}

exports.default = FileId;
