// @hashgraph/sdk/lib/contract/ContractId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _EntityId = _interopRequireWildcard(require("../EntityId.cjs"));

var proto = _interopRequireWildcard(require("@hashgraph/proto"));

var _long = _interopRequireDefault(require("long"));

var hex = _interopRequireWildcard(require("../encoding/hex.cjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * The ID for a crypto-currency contract on Hedera.
 *
 * @augments {EntityId<proto.IContractID>}
 */
class ContractId extends _EntityId.default {
  /**
   * @param {number | Long | import("../EntityId.js").IEntityId} properties
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(properties, realm, num) {
    super(properties, realm, num);
  }
  /**
   * @param {string} text
   * @returns {ContractId}
   */


  static fromString(text) {
    return new ContractId(...(0, _EntityId.fromString)(text));
  }
  /**
   * @internal
   * @param {proto.IContractID} id
   * @returns {ContractId}
   */


  static _fromProtobuf(id) {
    return new ContractId({
      shard: id.shardNum != null ? id.shardNum : 0,
      realm: id.realmNum != null ? id.realmNum : 0,
      num: id.contractNum != null ? id.contractNum : 0
    });
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {ContractId}
   */


  static fromBytes(bytes) {
    return ContractId._fromProtobuf(proto.ContractID.decode(bytes));
  }
  /**
   * @param {string} address
   * @returns {ContractId}
   */


  static fromSolidityAddress(address) {
    const addr = address.startsWith("0x") ? hex.decode(address.slice(2)) : hex.decode(address);

    if (addr.length !== 20) {
      throw new Error(`Invalid hex encoded solidity address length:
                    expected length 40, got length ${address.length}`);
    }

    const shard = _long.default.fromBytesBE(Array.from(addr.slice(0, 4)));

    const realm = _long.default.fromBytesBE(Array.from(addr.slice(4, 12)));

    const contract = _long.default.fromBytesBE(Array.from(addr.slice(12, 20)));

    return new ContractId(shard, realm, contract);
  }
  /**
   * @override
   * @internal
   * @returns {proto.IContractID}
   */


  _toProtobuf() {
    return {
      contractNum: this.num,
      shardNum: this.shard,
      realmNum: this.realm
    };
  }
  /**
   * @returns {Uint8Array}
   */


  toBytes() {
    return proto.ContractID.encode(this._toProtobuf()).finish();
  }

}

exports.default = ContractId;
