// @hashgraph/sdk/lib/token/TokenId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _EntityId = _interopRequireWildcard(require("../EntityId.cjs"));

var proto = _interopRequireWildcard(require("@hashgraph/proto"));

var _long = _interopRequireDefault(require("long"));

var hex = _interopRequireWildcard(require("../encoding/hex.cjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * The ID for a crypto-currency token on Hedera.
 *
 * @augments {EntityId<proto.ITokenID>}
 */
class TokenId extends _EntityId.default {
  /**
   * @param {number | Long | import("../EntityId.js").IEntityId} properties
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(properties, realm, num) {
    super(properties, realm, num);
  }
  /**
   * @param {string} text
   * @returns {TokenId}
   */


  static fromString(text) {
    return new TokenId(...(0, _EntityId.fromString)(text));
  }
  /**
   * @internal
   * @param {proto.ITokenID} id
   * @returns {TokenId}
   */


  static _fromProtobuf(id) {
    return new TokenId({
      shard: id.shardNum != null ? id.shardNum : 0,
      realm: id.realmNum != null ? id.realmNum : 0,
      num: id.tokenNum != null ? id.tokenNum : 0
    });
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {TokenId}
   */


  static fromBytes(bytes) {
    return TokenId._fromProtobuf(proto.TokenID.decode(bytes));
  }
  /**
   * @param {string} address
   * @returns {TokenId}
   */


  static fromSolidityAddress(address) {
    const addr = address.startsWith("0x") ? hex.decode(address.slice(2)) : hex.decode(address);

    if (addr.length !== 20) {
      throw new Error(`Invalid hex encoded solidity address length:
                    expected length 40, got length ${address.length}`);
    }

    const shard = _long.default.fromBytesBE(Array.from(addr.slice(0, 4)));

    const realm = _long.default.fromBytesBE(Array.from(addr.slice(4, 12)));

    const token = _long.default.fromBytesBE(Array.from(addr.slice(12, 20)));

    return new TokenId(shard, realm, token);
  }
  /**
   * @internal
   * @override
   * @returns {proto.ITokenID}
   */


  _toProtobuf() {
    return {
      tokenNum: this.num,
      shardNum: this.shard,
      realmNum: this.realm
    };
  }
  /**
   * @returns {Uint8Array}
   */


  toBytes() {
    return proto.TokenID.encode(this._toProtobuf()).finish();
  }

}

exports.default = TokenId;
