// @hashgraph/sdk/lib/token/TokenUpdateTransaction.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _TokenId = _interopRequireDefault(require("./TokenId.cjs"));

var _Transaction = _interopRequireWildcard(require("../transaction/Transaction.cjs"));

var _protobuf = require("../cryptography/protobuf.cjs");

var _long = _interopRequireDefault(require("long"));

var _AccountId = _interopRequireDefault(require("../account/AccountId.cjs"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").ITransaction} proto.ITransaction
 * @typedef {import("@hashgraph/proto").TransactionBody} proto.TransactionBody
 * @typedef {import("@hashgraph/proto").ITransactionBody} proto.ITransactionBody
 * @typedef {import("@hashgraph/proto").ITransactionResponse} proto.ITransactionResponse
 * @typedef {import("@hashgraph/proto").ITokenUpdateTransactionBody} proto.ITokenUpdateTransactionBody
 * @typedef {import("@hashgraph/proto").ITokenID} proto.ITokenID
 */

/**
 * @typedef {import("bignumber.js").default} BigNumber
 * @typedef {import("@hashgraph/cryptography").Key} Key
 * @typedef {import("../channel/Channel.js").default} Channel
 */

/**
 * Update a new Hedera™ crypto-currency token.
 */
class TokenUpdateTransaction extends _Transaction.default {
  /**
   * @param {object} [props]
   * @param {TokenId | string} [props.tokenId]
   * @param {string} [props.name]
   * @param {string} [props.symbol]
   * @param {AccountId | string} [props.treasury]
   * @param {Key} [props.adminKey]
   * @param {Key} [props.kycKey]
   * @param {Key} [props.freezeKey]
   * @param {Key} [props.wipeKey]
   * @param {Key} [props.supplyKey]
   * @param {AccountId | string} [props.autoRenewAccount]
   * @param {Long | number} [props.expirationTime]
   * @param {Long | number} [props.autoRenewPeriod]
   */
  constructor(props = {}) {
    super();
    /**
     * @private
     * @type {?TokenId}
     */

    this._tokenId = null;
    /**
     * @private
     * @type {?string}
     */

    this._name = null;
    /**
     * @private
     * @type {?string}
     */

    this._symbol = null;
    /**
     * @private
     * @type {?AccountId}
     */

    this._treasury = null;
    /**
     * @private
     * @type {?Key}
     */

    this._adminKey = null;
    /**
     * @private
     * @type {?Key}
     */

    this._kycKey = null;
    /**
     * @private
     * @type {?Key}
     */

    this._freezeKey = null;
    /**
     * @private
     * @type {?Key}
     */

    this._wipeKey = null;
    /**
     * @private
     * @type {?Key}
     */

    this._supplyKey = null;
    /**
     * @private
     * @type {?AccountId}
     */

    this._autoRenewAccount = null;
    /**
     * @private
     * @type {?Long}
     */

    this._expirationTime = null;
    /**
     * @private
     * @type {?Long}
     */

    this._autoRenewPeriod = null;

    if (props.tokenId != null) {
      this.setTokenId(props.tokenId);
    }

    if (props.name != null) {
      this.setName(props.name);
    }

    if (props.symbol != null) {
      this.setSymbol(props.symbol);
    }

    if (props.treasury != null) {
      this.setTreasury(props.treasury);
    }

    if (props.adminKey != null) {
      this.setAdminKey(props.adminKey);
    }

    if (props.kycKey != null) {
      this.setKycKey(props.kycKey);
    }

    if (props.freezeKey != null) {
      this.setFreezeKey(props.freezeKey);
    }

    if (props.wipeKey != null) {
      this.setWipeKey(props.wipeKey);
    }

    if (props.supplyKey != null) {
      this.setsupplyKey(props.supplyKey);
    }

    if (props.autoRenewAccount != null) {
      this.setAutoRenewAccount(props.autoRenewAccount);
    }

    if (props.expirationTime != null) {
      this.setExpirationTime(props.expirationTime);
    }

    if (props.autoRenewPeriod != null) {
      this.setAutoRenewPeriod(props.autoRenewPeriod);
    }
  }
  /**
   * @internal
   * @param {proto.ITransactionBody} body
   * @returns {TokenUpdateTransaction}
   */


  static _fromProtobuf(body) {
    const update =
    /** @type {proto.ITokenUpdateTransactionBody} */
    body.tokenCreation;
    return new TokenUpdateTransaction({
      tokenId: update.token != null ? _TokenId.default._fromProtobuf(update.token) : undefined,
      name: update.name != null ? update.name : undefined,
      symbol: update.symbol != null ? update.symbol : undefined,
      treasury: update.treasury != null ? _AccountId.default._fromProtobuf(update.treasury) : undefined,
      adminKey: update.adminKey != null ? (0, _protobuf.keyFromProtobuf)(update.adminKey) : undefined,
      kycKey: update.kycKey != null ? (0, _protobuf.keyFromProtobuf)(update.kycKey) : undefined,
      freezeKey: update.freezeKey != null ? (0, _protobuf.keyFromProtobuf)(update.freezeKey) : undefined,
      wipeKey: update.wipeKey != null ? (0, _protobuf.keyFromProtobuf)(update.wipeKey) : undefined,
      supplyKey: update.supplyKey != null ? (0, _protobuf.keyFromProtobuf)(update.supplyKey) : undefined,
      autoRenewAccount: update.autoRenewAccount != null ? _AccountId.default._fromProtobuf(update.autoRenewAccount) : undefined,
      expirationTime: update.expiry != null ? update.expiry : undefined,
      autoRenewPeriod: update.autoRenewPeriod != null ? update.autoRenewPeriod : undefined
    });
  }
  /**
   * @returns {?TokenId}
   */


  get tokenId() {
    return this._tokenId;
  }
  /**
   * @param {TokenId | string} tokenId
   * @returns {this}
   */


  setTokenId(tokenId) {
    this._requireNotFrozen();

    this._tokenId = tokenId instanceof _TokenId.default ? tokenId : _TokenId.default.fromString(tokenId);
    return this;
  }
  /**
   * @returns {?string}
   */


  get name() {
    return this._name;
  }
  /**
   * @param {string} name
   * @returns {this}
   */


  setName(name) {
    this._requireNotFrozen();

    this._name = name;
    return this;
  }
  /**
   * @returns {?string}
   */


  get symbol() {
    return this._symbol;
  }
  /**
   * @param {string} symbol
   * @returns {this}
   */


  setSymbol(symbol) {
    this._requireNotFrozen();

    this._symbol = symbol;
    return this;
  }
  /**
   * @returns {?AccountId}
   */


  get treasury() {
    return this._treasury;
  }
  /**
   * @param {AccountId | string} id
   * @returns {this}
   */


  setTreasury(id) {
    this._requireNotFrozen();

    this._treasury = id instanceof _AccountId.default ? id : _AccountId.default.fromString(id);
    return this;
  }
  /**
   * @returns {?Key}
   */


  get adminKey() {
    return this._adminKey;
  }
  /**
   * @param {Key} key
   * @returns {this}
   */


  setAdminKey(key) {
    this._requireNotFrozen();

    this._adminKey = key;
    return this;
  }
  /**
   * @returns {?Key}
   */


  get kycKey() {
    return this._kycKey;
  }
  /**
   * @param {Key} key
   * @returns {this}
   */


  setKycKey(key) {
    this._requireNotFrozen();

    this._kycKey = key;
    return this;
  }
  /**
   * @returns {?Key}
   */


  get freezeKey() {
    return this._freezeKey;
  }
  /**
   * @param {Key} key
   * @returns {this}
   */


  setFreezeKey(key) {
    this._requireNotFrozen();

    this._freezeKey = key;
    return this;
  }
  /**
   * @returns {?Key}
   */


  get wipeKey() {
    return this._wipeKey;
  }
  /**
   * @param {Key} key
   * @returns {this}
   */


  setWipeKey(key) {
    this._requireNotFrozen();

    this._wipeKey = key;
    return this;
  }
  /**
   * @returns {?Key}
   */


  get supplyKey() {
    return this._supplyKey;
  }
  /**
   * @param {Key} key
   * @returns {this}
   */


  setsupplyKey(key) {
    this._requireNotFrozen();

    this._supplyKey = key;
    return this;
  }
  /**
   * @returns {?Long}
   */


  get expirationTime() {
    return this._expirationTime;
  }
  /**
   * @param {Long | number} time
   * @returns {this}
   */


  setExpirationTime(time) {
    this._requireNotFrozen();

    this._expirationTime = time instanceof _long.default ? time : _long.default.fromValue(time);
    return this;
  }
  /**
   * @returns {?AccountId}
   */


  get autoRenewAccount() {
    return this._autoRenewAccount;
  }
  /**
   * @param {AccountId | string} id
   * @returns {this}
   */


  setAutoRenewAccount(id) {
    this._requireNotFrozen();

    this._autoRenewAccount = id instanceof _AccountId.default ? id : _AccountId.default.fromString(id);
    return this;
  }
  /**
   * @returns {?Long}
   */


  get autoRenewPeriod() {
    return this._autoRenewPeriod;
  }
  /**
   * Set the auto renew period for this token.
   *
   * @param {number | Long} autoRenewPeriod
   * @returns {this}
   */


  setAutoRenewPeriod(autoRenewPeriod) {
    this._requireNotFrozen();

    this._autoRenewPeriod = autoRenewPeriod instanceof _long.default ? autoRenewPeriod : _long.default.fromValue(autoRenewPeriod);
    return this;
  }
  /**
   * @override
   * @internal
   * @param {Channel} channel
   * @param {proto.ITransaction} request
   * @returns {Promise<proto.ITransactionResponse>}
   */


  _execute(channel, request) {
    return channel.token.updateToken(request);
  }
  /**
   * @override
   * @protected
   * @returns {NonNullable<proto.TransactionBody["data"]>}
   */


  _getTransactionDataCase() {
    return "tokenUpdate";
  }
  /**
   * @override
   * @protected
   * @returns {proto.ITokenUpdateTransactionBody}
   */


  _makeTransactionData() {
    return {
      token: this._tokenId != null ? this._tokenId._toProtobuf() : null,
      name: this.name,
      symbol: this.symbol,
      treasury: this._treasury != null ? this._treasury._toProtobuf() : null,
      adminKey: this._adminKey != null ? (0, _protobuf.keyToProtobuf)(this._adminKey) : null,
      kycKey: this._kycKey != null ? (0, _protobuf.keyToProtobuf)(this._kycKey) : null,
      freezeKey: this._freezeKey != null ? (0, _protobuf.keyToProtobuf)(this._freezeKey) : null,
      wipeKey: this._wipeKey != null ? (0, _protobuf.keyToProtobuf)(this._wipeKey) : null,
      supplyKey: this._supplyKey != null ? (0, _protobuf.keyToProtobuf)(this._supplyKey) : null,
      autoRenewAccount: this._autoRenewAccount != null ? this._autoRenewAccount._toProtobuf() : null,
      expiry: this._expirationTime,
      autoRenewPeriod: this._autoRenewPeriod
    };
  }

}

exports.default = TokenUpdateTransaction;

_Transaction.TRANSACTION_REGISTRY.set("tokenUpdate", // eslint-disable-next-line @typescript-eslint/unbound-method
TokenUpdateTransaction._fromProtobuf);
